### PR TITLE
Remove identifiable from document::DataType

### PR DESCRIPTION
**Pull request by baldersheim**
Also remove the virtual == operator and the need for default constructor.

@vekterli or @toregge PR

### CODE REVIEW

File: streamingvisitors/src/vespa/searchvisitor/searchvisitor.cpp [hunk 1]
```diff
@@ -739,12 +740,12 @@ void SearchVisitor::setupAttributeVector(const FieldPath &fieldPath) {
     enum FieldDataType { OTHER = 0, ARRAY, WSET };
     FieldDataType typeSeen = OTHER;
     for (const auto & entry : fieldPath) {
-        int dataTypeId(entry->getDataType().getClass().id());
-        if (dataTypeId == document::ArrayDataType::classId) {
+        const document::DataType & dt = entry->getDataType();
+        if (dt.isArray()) {
             typeSeen = ARRAY;
-        } else if (dataTypeId == document::MapDataType::classId) {
+        } else if (dt.cast_map() != nullptr) {
```
Comment (vekterli) [inline streamingvisitors/src/vespa/searchvisitor/searchvisitor.cpp:746]: Should there be a distinct `isMap()` for symmetry?

**Review by vekterli — APPROVED** [review]
:+1: woohoo 🥳

**Review by toregge — APPROVED** [review]

File: document/src/vespa/document/datatype/arraydatatype.h [hunk 2]
```diff
@@ -12,21 +12,16 @@
...
-protected:
-    // Protected to help you avoid calling the copy constructor when
-    // you think you're calling the regular constructor with a nested
-    // ArrayDataType.
-    ArrayDataType(const ArrayDataType &o) : CollectionDataType(o) {}
-
 public:
     explicit ArrayDataType(const DataType &nestedType);
+    ArrayDataType(const ArrayDataType &o) = delete;
+    ArrayDataType &operator=(const ArrayDataType &rhs) = delete;
```
Comment (toregge) [inline document/src/vespa/document/datatype/arraydatatype.h:18]: Consider explicitly deleting move constructor and move assignment too.

**Review by toregge — APPROVED** [review]

File: streamingvisitors/src/vespa/searchvisitor/searchvisitor.cpp (hunk 1 above)
Comment (baldersheim) [reply streamingvisitors/src/vespa/searchvisitor/searchvisitor.cpp]: Fixed

File: document/src/vespa/document/datatype/arraydatatype.h (hunk 2 above)
Comment (baldersheim) [reply document/src/vespa/document/datatype/arraydatatype.h:18]: I will do another pass cleaning up the todos and also delete the move constructor/operator.
But for they are deleted implicit by the explicit destructor.

**Pull request merged**